File: yami4/src/tools/yami4idl/examples/java/subscription/Publisher.java
import com.inspirel.yami.Agent;
import com.inspirel.yami.Parameters;
import com.inspirel.yami.ValuePublisher;

import static java.lang.Math.abs;
import java.util.Random;

public class Publisher {

    public static void main(String[] args) {
        if (args.length != 1) {
            System.out.println(
                "expecting one parameter: " +
                "publisher destination");
            return;
        }

        String publisherAddress = args[0];

        try {
            ValuePublisher randomValue =
                new ValuePublisher();

            Agent publisherAgent = new Agent();

            String resolvedAddress =
                publisherAgent.addListener(publisherAddress);

            System.out.println(
                "The publisher is listening on " +
                resolvedAddress);

            publisherAgent.registerValuePublisher(
                "random_number", randomValue);

            // publish random values forever
            Parameters content = new Parameters();
            Random generator = new Random();
            while (true) {
                subscription.Payload p =
                    new subscription.Payload();
                    
                int random =
                    abs(generator.nextInt()) % 100;
                p.value = random;

                System.out.println(
                    "publishing value " + random);

                Parameters params = new Parameters();
                p.write(params);
                
                randomValue.publish(params);

                Thread.sleep(1000);
            }
        } catch (Exception ex) {
            System.out.println(
                "error: " + ex.getMessage());
        }
    }
}

File: yami4/src/tools/yami4idl/examples/cpp/subscription/publisher.cpp
#include "subscription.h"
#include <yami4-cpp/yami.h>

#include <cstdlib>
#include <iostream>

#include <pause.h>

int main(int argc, char * argv[])
{
    if (argc != 2)
    {
        std::cout
            << "expecting one parameter: "
            << "publisher destination\n";
        return EXIT_FAILURE;
    }

    const std::string publisher_address = argv[1];

    try
    {
        yami::value_publisher random_value;

        yami::agent publisher_agent;

        const std::string resolved_address =
            publisher_agent.add_listener(publisher_address);

        std::cout << "The publisher is listening on "
            << resolved_address << std::endl;

        publisher_agent.register_value_publisher(
            "random_number", random_value);

        // publish random numbers forever
        while (true)
        {
            subscription::payload p;
            
            const int random = std::rand() % 100;
            p.value = random;

            std::cout
                << "publishing value " << random << std::endl;

            yami::parameters params;
            p.write(params);
            
            random_value.publish(params);

            // pause for 1s
            examples::pause();
        }
    }
    catch (const std::exception & e)
    {
        std::cout << "error: " << e.what() << std::endl;
    }
}
